backend/postgres/postgres.go

```diff
@@ -0,0 +1,856 @@
+
+	if options.ApplyMigrations {
+		if err := b.Migrate(); err != nil {
+			panic(err)
+		}
+	}
+
+	return b
+}
+
+type postgresBackend struct {
+	dsn        string
+	db         *sql.DB
+	workerName string
+	options    *options
+}
+
+func (mb *postgresBackend) Close() error {
+	return mb.db.Close()
+}
+
+// Migrate applies any pending database migrations.
+func (mb *postgresBackend) Migrate() error {
+	schemaDsn := mb.dsn
+	db, err := sql.Open("pgx", schemaDsn)
+	if err != nil {
+		return fmt.Errorf("opening schema database: %w", err)
+	}
+
+	dbi, err := postgres.WithInstance(db, &postgres.Config{})
+	if err != nil {
+		return fmt.Errorf("creating migration instance: %w", err)
+	}
+
+	migrations, err := iofs.New(migrationsFS, "db/migrations")
+	if err != nil {
+		return fmt.Errorf("creating migration source: %w", err)
+	}
+
+	m, err := migrate.NewWithInstance("iofs", migrations, "postgres", dbi)
+	if err != nil {
+		return fmt.Errorf("creating migration: %w", err)
+	}
+
+	if err := m.Up(); err != nil {
+		if !errors.Is(err, migrate.ErrNoChange) {
+			return fmt.Errorf("running migrations: %w", err)
+		}
+	}
+
+	if err := db.Close(); err != nil {
+		return fmt.Errorf("closing schema database: %w", err)
+	}
+
+	return nil
+}
+
+func (b *postgresBackend) Logger() *slog.Logger {
+	return b.options.Logger
+}
+
+func (b *postgresBackend) Tracer() trace.Tracer {
+	return b.options.TracerProvider.Tracer(backend.TracerName)
+}
+
+func (b *postgresBackend) Metrics() metrics.Client {
+	return b.options.Metrics.WithTags(metrics.Tags{metrickeys.Backend: "postgres"})
+}
+
+func (b *postgresBackend) Converter() converter.Converter {
+	return b.options.Converter
+}
+
+func (b *postgresBackend) ContextPropagators() []workflow.ContextPropagator {
+	return b.options.ContextPropagators
+}
+
+func (b *postgresBackend) CreateWorkflowInstance(ctx context.Context, instance *workflow.Instance, event *history.Event) error {
+	tx, err := b.db.BeginTx(ctx, &sql.TxOptions{
+		Isolation: sql.LevelReadCommitted,
+	})
+	if err != nil {
+		return fmt.Errorf("starting transaction: %w", err)
+	}
+	defer tx.Rollback()
+
+	// Create workflow instance
+	if err := createInstance(ctx, tx, instance, event.Attributes.(*history.ExecutionStartedAttributes).Metadata); err != nil {
+		return err
+	}
+
+	// Initial history is empty, store only new events
+	if err := insertPendingEvents(ctx, tx, instance, []*history.Event{event}); err != nil {
+		return fmt.Errorf("inserting new event: %w", err)
+	}
+
+	if err := tx.Commit(); err != nil {
+		return fmt.Errorf("creating workflow instance: %w", err)
+	}
+
+	return nil
+}
+
+func (b *postgresBackend) RemoveWorkflowInstance(ctx context.Context, instance *core.WorkflowInstance) error {
+	tx, err := b.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	row := tx.QueryRowContext(ctx, SQLReplacer("SELECT state FROM instances WHERE instance_id = ? AND execution_id = ? LIMIT 1"), instance.InstanceID, instance.ExecutionID)
```

Comment:
@erma07 I think you are fine to use positional parameters when using pgx, e.g.

```suggestion
	row := tx.QueryRowContext(
	  ctx, 
	  "SELECT state FROM instances WHERE instance_id = $1 AND execution_id = $2 LIMIT 1",
	  instance.InstanceID, 
	  instance.ExecutionID,
	)
```